### PR TITLE
Raise TraitNotificationError on trailing comma in on_trait_change spec.

### LINKED ISSUE #406: on_trait_change: trailing comma in an extended name causes surprises
Decorating a `HasTraits` subclass method with `@on_trait_change('foo,bar,')` has the surprising side-effect of listening to everything. In most parts of Python, trailing commas are innocuous.

Example script:
```python
from traits.api import *

class A(HasTraits):
    foo = Int()
    bar = Int()
    baz = Int()

    @on_trait_change('foo,bar,')
    def _report_change(self, obj, name, old, new):
        print("{!r} trait changed from {!r} to {!r}".format(name, old, new))


a = A()
a.bar = 5  # expect a single notification; get two notifications
a.baz = 3  # expect no notification, but a notification occurs
```

On running this, I see the following output:
```
bash-4.4$ python bug.py 
'bar' trait changed from 0 to 5
'bar' trait changed from 0 to 5
'baz' trait changed from 0 to 3
```

With the trailing comma removed, I get:
```
bash-4.4$ python bug.py 
'bar' trait changed from 0 to 5
```

Labelling as a bug: even if this is working as intended behaviour (which seems possible), it's at least a usability bug, and I've seen it trip up newcomers on more than one occasion. I think we should fix this, even at the risk of possibly breaking existing code.

The fix I'd propose is making a trailing comma here an error: this would give a noisy early breakage for any projects that are intentionally or accidentally using this pattern, rather than the subtle late-discovered breakage that might occur if we simply allow the trailing comma but change the behaviour.

### PR DESCRIPTION
**Pull request by midhun-pm**
Fixes #406 

Modifies the `ListenerParser` which parses the `name` for the `on_trait_change`decorator to check for a trailing `,`.
Raises a `TraitNotificationError` error if trailing comma is present in the trimmed string.

### CODE REVIEW

File: traits/tests/test_on_trait_change.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+class BaseTestClass(HasTraits):
+    foo = Int()
+    bar = Int()
+    baz = Int()
+    output_name = None
+    output_old = None
+    output_new = None
+
+
+class TestOnTraitChangeNotification(unittest.TestCase):
```
Comment (corranwebster) [inline traits/tests/test_on_trait_change.py]: This should probably be in `test_extended_trait_change` - that is where all the testing of the different listening patterns is, I think.

File: traits/traits_listener.py [hunk 2]
```diff
@@ -1024,6 +1024,10 @@ def parse(self):
         # a shortcut too. Whitespace should already have been stripped from the
         # start and end.
 
+        if self.text.strip().endswith(","):
+            raise TraitNotificationError("Error parsing name. "
```
Comment (kitchoi) [inline traits/traits_listener.py]: Please use `self.error` instead in order to raise an exception type consistent with `parse_group`.

**Comment by midhun-pm** [comment]
Addressed PR comments. Sorry for the delay, lost track of this one.

File: traits/tests/test_extended_trait_change.py [hunk 3]
```diff
@@ -117,6 +117,13 @@ def arg_check4(self, object, name, old, new):
         self.tc.assertEqual(new, self.value)
 
 
+class ArgCheckDecoratorTrailingComma(ArgCheckDecorator):
+    @on_trait_change("int1, int2,")
+    def arg_check(self, object, name, old, new):
+        self.calls += 1
+        self.tc.assertIs(object, self)
```
Comment (kitchoi) [inline traits/tests/test_extended_trait_change.py]: Nit: These two lines can probably be replaced by `pass`.

**Comment by mdickinson** [comment]
The macOS Python 3.6 build seems flaky today. I've restarted it (again).

**Comment by mdickinson** [comment]
Ah, sorry; forgot that my own previous review would block merging.

**Pull request merged**